lib/Dialect/TensorExt/Transforms/ImplementShiftNetwork.cpp

```diff
@@ -0,0 +1,409 @@
+    arith::MulIOp maskOp =
+        rewriter.create<arith::MulIOp>(tensor.getLoc(), tensor, mask);
+    Value rotated = rewriter.create<tensor_ext::RotateOp>(
+        tensor.getLoc(), maskOp.getResult(),
+        rewriter.create<arith::ConstantIntOp>(tensor.getLoc(), rotationAmount,
+                                              rewriter.getI32Type()));
+
+    if (result.has_value()) {
+      result = rewriter.create<arith::AddIOp>(tensor.getLoc(), result.value(),
+                                              rotated);
+    } else {
+      result = rotated;
+    }
+  }
+
+  return result.has_value() ? result.value() : tensor;
+}
+
+LogicalResult convertLayoutOp(PermuteOp op,
+                              VosVosErkinShiftNetworks &shiftNetworks,
+                              int64_t ciphertextSize) {
+  LLVM_DEBUG(llvm::dbgs() << "Converting layout op: " << op << "\n");
+  IRRewriter rewriter(op.getContext());
+  ImplicitLocOpBuilder b(op.getLoc(), rewriter);
+  RankedTensorType tensorTy = op.getInput().getType();
+
+  // Only support a 1-D tensor until sharding is supported
+  if (op.getInput().getType().getRank() != 1) {
+    return op.emitError("requires a one-dimensional tensor");
+  }
+
+  SmallVector<int64_t> permutation;
+
+  // Convert the affine map to an explicit permutation, or else use the dense
+  // array attr as an explicit permutation.
+  auto affineMapAttr = dyn_cast<AffineMapAttr>(op.getPermutation());
+  if (affineMapAttr) {
+    AffineMap permutationMap = simplifyAffineMap(affineMapAttr.getValue());
+
+    if (permutationMap.isIdentity()) {
+      rewriter.replaceOp(op, op.getOperand());
+      return success();
+    }
+
+    LLVM_DEBUG(llvm::dbgs()
+               << "Expanding permutation from " << permutationMap << "\n");
+    if (failed(makeExplicit1DMapping(permutationMap, tensorTy.getNumElements(),
+                                     permutation)))
+      return failure();
+  } else {
+    auto denseElementsAttr =
+        dyn_cast<DenseIntElementsAttr>(op.getPermutation());
+    if (denseElementsAttr) {
+      permutation = llvm::map_to_vector(
+          denseElementsAttr, [](const APInt &i) { return i.getSExtValue(); });
+    } else {
+      return failure();
+    }
+  }
+
+  if (!isPermutation(permutation)) {
+    auto diag = op.emitError(
+        "expected a permutation, but got a mapping that was not a "
+        "permutation.");
+    auto &note = diag.attachNote() << "mapping was:\n";
+    for (int i = 0; i < permutation.size(); i++) {
+      note << i << " -> " << permutation[i] << ", ";
+      if (i % 10 == 9) {
+        note << "\n";
+      }
+    }
+    return diag;
+  }
+
+  LLVM_DEBUG({
+    llvm::dbgs() << "PermuteOp produces underlying permutation: ";
+    for (int i = 0; i < permutation.size(); i++) {
+      llvm::dbgs() << i << " -> " << permutation[i] << ", ";
+      if (i % 10 == 9) {
+        llvm::dbgs() << "\n";
+      }
+    }
+    llvm::dbgs() << "\n";
```

Comment:
maybe you can have a printer for the permutation to reuse here and above?

Reply:
Done